### ISSUE TITLE
Inconsistent behavior when introspecting parameterized types

### ISSUE THREAD

**Issue by morozov** [vimeo/psalm#5573]
https://psalm.dev/r/8137eb841a

As you can see in the example, the `D1` and `D2` types have absolutely the same structure but checking `instaceof D1` is considered value while checking `! instaceof D2` is considered redundant.

Both cases look valid but if they aren't, they should be consistently marked as redundant.

**Comment by morozov** [comment]
Hi, any chance for this to get triaged? Thanks!

**Comment by weirdan** [comment]
Simplified: https://psalm.dev/r/8558e22043

I believe both should be flagged (positive case with something like `ParadoxicalCondition` or `TypeDoesNotContainType` and negative with `RedundantCondition`), because P is an invariant type parameter. Indeed with invariant template `Driver<Throwable>` cannot be passed `D1` which is `Driver<Error>`.

**Comment by morozov** [comment]
Thanks, @weirdan! I saw the documentation on type variance but wasn't sure if it's my case.

**Comment by morozov** [comment]
The syntax of generics in Psalm looks very similar to the syntax in Java but it looks like the semantics are different.

```java
public interface Platform {
}

public class D1 implements Driver<P1> {
}

public class D2 implements Driver<P2> {
}

public interface Driver<P extends Platform> {
}

public class P1 implements Platform {
}

public class P2 implements Platform {
}

public class Test {
    public Test(Driver<?> driver) {
        if (driver instanceof D1) {
            System.out.println("D1");
        }

        if (driver instanceof D2) {
            System.out.println("D2");
        }
    }
}
```
Above is a valid Java code example which I want to replicate in PHP with Psalm. I want the test method to accept a `Driver` parametrized with any type within the type constraint (which in the Java example above is expressed via `<?>`). I don't want the parameter to be covariant.

How can I express that in Psalm?

**Comment by morozov** [comment]
Another constraint I'd like to express with Psalm is the following:
```java
public void acceptDriverClass(Class<? extends Driver<?>> driverClass) {
}

public void testD1() {
    acceptDriverClass(D1.class);
}
```
The function needs to accept a class that implements the Driver interface and parameterized with anything.

How do I achieve this with Psalm? This is a boilerplate code: https://psalm.dev/r/8a510ef143. Psalm is expected to allow `testD1()` but flag `testP1()`.

**Comment by AndrolGenhald** [comment]
Psalm is expected to allow testD1() but flag testP1().

That can be achieved with `@template-covariant`: https://psalm.dev/r/2c32e48b07

Eventually I'd like to support `T of Type`, `T extends Type`, and `T super Type` for invariance, covariance, and contravariance respectively, but wading through all of the template code hasn't been high on my priority list, and it's likely to be a lot of work if/when I ever get around to it.
